File: ios/SiriApp/Chat/ChatSession.swift
//
//  ChatSession.swift
//  SiriApp
//
//  Multi-turn conversation manager. Keeps last MAX_HISTORY messages.
//  Supports RAG (Retrieval-Augmented Generation) via HybridSearcher.
//
//  Ported from Android: ChatSession.kt
//

import Foundation
import Combine
import os.log

class ChatSession: ObservableObject {
    /// Messages shown on screen — capped at maxScreenMessages.
    @Published private(set) var messages: [ChatMessage] = []

    /// LLM context buffer — preserved across screen clears so the model
    /// remembers previous turns. Never cleared except by user action.
    private var contextBuffer: [ChatMessage] = []

    /// Maximum messages shown on screen (older messages are trimmed).
    private let maxScreenMessages: Int

    /// Context window for LLM — last N messages from the full context buffer.
    private var contextMessages: [ChatMessage] {
        Array(contextBuffer.suffix(maxHistory))
    }

    private let llmClient: LlmClient
    private let configRepository: ConfigRepository
    private let hybridSearcher: HybridSearcher?
    private let maxHistory: Int
    private var cancellables = Set<AnyCancellable>()

    private let log = OSLog(subsystem: "dev.richard.voicechat", category: "ChatSession")

    init(llmClient: LlmClient,
         configRepository: ConfigRepository,
         hybridSearcher: HybridSearcher? = nil,
         maxHistory: Int = 5,
         maxScreenMessages: Int = 20) {
        self.llmClient = llmClient
        self.configRepository = configRepository
        self.hybridSearcher = hybridSearcher
        self.maxHistory = maxHistory
        self.maxScreenMessages = maxScreenMessages
    }

    // MARK: - RAG Context Retrieval

    /// 混合检索：向量语义 + BM25 关键词 → RRF 融合。
    /// 如果 HybridSearcher 未配置或 RAG 被禁用，返回 nil。
    private func retrieveContext(_ userText: String) async -> String? {
        guard let hybridSearcher = hybridSearcher else { return nil }
        guard configRepository.isRagEnabled() else { return nil }

        let results = await hybridSearcher.search(query: userText, topK: 3)
        if results.isEmpty {
            os_log(.debug, log: log, "retrieveContext: no relevant chunks found")
            return nil
        }

        os_log(.info, log: log,
               "retrieveContext: found %{public}d chunks, top RRF scores: %{public}@",
               results.count,
               results.map { String(format: "%.4f", $0.score) }.joined(separator: ", "))

        return results.map { $0.content }.joined(separator: "\n---\n")
    }

    // MARK: - Send

    /// Send message (non-streaming)
    func send(_ text: String) -> AnyPublisher<String, Error> {
        let userMsg = ChatMessage(role: .user, content: text)
        appendToScreen(userMsg)
        contextBuffer.append(userMsg)

        return Future<String, Error> { [weak self] promise in
            guard let self = self else { return }
            Task {
                let ragContext = await self.retrieveContext(text)
                let publisher = self.llmClient.chat(
                    messages: self.contextMessages,
                    ragContext: ragContext
                )
                var cancellable: AnyCancellable?
                var didFulfill = false
                cancellable = publisher.sink(
                    receiveCompletion: { completion in
                        if case .failure(let error) = completion {
                            self.messages.removeLast()
                            self.contextBuffer.removeLast()
                            if !didFulfill { didFulfill = true; promise(.failure(error)) }
                        }
                        _ = cancellable
                    },
                    receiveValue: { reply in
                        let assistantMsg = ChatMessage(role: .assistant, content: reply)
                        self.appendToScreen(assistantMsg)
                        self.contextBuffer.append(assistantMsg)
                        if !didFulfill { didFulfill = true; promise(.success(reply)) }
                    }
                )
            }
        }.eraseToAnyPublisher()
    }

    /// Send message with streaming (iOS 14+).
    /// Retrieves RAG context asynchronously before starting the LLM stream.
    func sendStream(_ text: String) -> AnyPublisher<AnyPublisher<String, Error>, Error> {
        let userMsg = ChatMessage(role: .user, content: text)
        appendToScreen(userMsg)
        contextBuffer.append(userMsg)

        // If RAG is not available, skip the async wrapper for lower latency
        guard hybridSearcher != nil && configRepository.isRagEnabled() else {
            let streamPublisher = llmClient.chatStreamPublisher(messages: contextMessages)
            return Just(streamPublisher)
                .setFailureType(to: Error.self)
                .eraseToAnyPublisher()
        }

        // Wrap the async retrieval in a Deferred+Future so the outer publisher
        // emits the inner stream publisher once RAG context is ready.
        return Deferred {
            Future<AnyPublisher<String, Error>, Error> { [weak self] promise in
                guard let self = self else { return }

                Task { [weak self] in
                    guard let self = self else { return }
                    let ragContext = await self.retrieveContext(text)
                    let streamPublisher = self.llmClient.chatStreamPublisher(
                        messages: self.contextMessages,
                        ragContext: ragContext
                    )
                    promise(.success(streamPublisher))
                }
            }
        }.eraseToAnyPublisher()
    }

    /// Save assistant reply to history (called after streaming completes)
    func appendAssistantReply(_ text: String) {
        guard text.isNotBlank else { return }
        let msg = ChatMessage(role: .assistant, content: text)
        appendToScreen(msg)
        contextBuffer.append(msg)
    }

    // MARK: - Screen / Context Management

    /// Append a message to the on-screen list, trimming to maxScreenMessages.
    private func appendToScreen(_ msg: ChatMessage) {
        messages.append(msg)
        if messages.count > maxScreenMessages {
            messages = Array(messages.suffix(maxScreenMessages))
        }
    }

    /// Clear screen only — LLM context is preserved.
    func clearScreen() {
        messages = []
    }

    /// Full clear — both screen and LLM context (user-initiated).
    func clear() {
        messages = []
        contextBuffer = []
    }

    var messageCount: Int {
        messages.count
    }
}

File: android/app/src/main/java/com/rd/siri/chat/ChatSession.kt
package com.rd.siri.chat

import android.util.Log
import com.rd.siri.config.ConfigRepository
import com.rd.siri.model.ChatMessage
import com.rd.siri.rag.HybridSearcher
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class ChatSession(
    private val llmClient: LlmClient,
    private val configRepository: ConfigRepository,
    private val hybridSearcher: HybridSearcher? = null,
    private val maxHistory: Int = 5,
    private val maxScreenMessages: Int = 20,
    private val maxContextBufferSize: Int = 200
) {

    companion object {
        private const val TAG = "SiriApp"
    }

    /** Messages shown on screen — capped at [maxScreenMessages]. */
    private val _messages = MutableStateFlow<List<ChatMessage>>(emptyList())
    val messages: StateFlow<List<ChatMessage>> = _messages.asStateFlow()

    /**
     * Full LLM context buffer — preserved across screen clears.
     * Grows unboundedly; only the last [maxHistory] messages are sent to the LLM.
     */
    private val contextBuffer = mutableListOf<ChatMessage>()

    /** LLM context window: last [maxHistory] messages from the full context buffer. */
    private val contextMessages: List<ChatMessage>
        get() = contextBuffer.takeLast(maxHistory)

    /** Append a message to the on-screen list, trimming to [maxScreenMessages]. */
    private fun appendToScreen(msg: ChatMessage) {
        _messages.value = (_messages.value + msg).takeLast(maxScreenMessages)
    }

    /** Append to context buffer and trim to [maxContextBufferSize] to bound memory. */
    private fun appendToContext(msg: ChatMessage) {
        contextBuffer.add(msg)
        if (contextBuffer.size > maxContextBufferSize) {
            val excess = contextBuffer.size - maxContextBufferSize
            contextBuffer.subList(0, excess).clear()
        }
    }

    /**
     * 混合检索：向量语义 + BM25 关键词 → RRF 融合。
     * 如果 HybridSearcher 未配置或 RAG 被禁用，返回 null。
     */
    private suspend fun retrieveContext(userText: String): String? {
        if (hybridSearcher == null) return null
        if (!configRepository.isRagEnabled()) return null

        val results = hybridSearcher.search(userText, topK = 3)
        if (results.isEmpty()) {
            Log.d(TAG, "retrieveContext: no relevant chunks found")
            return null
        }

        Log.i(TAG, "retrieveContext: found ${results.size} chunks, " +
                "top RRF scores: " + results.map { "%.4f".format(it.score) })

        return results.joinToString("\n---\n") { result ->
            result.content
        }
    }

    suspend fun send(text: String): Result<String> {
        val userMsg = ChatMessage(role = ChatMessage.Role.USER, content = text)
        appendToScreen(userMsg)
        appendToContext(userMsg)

        val ragContext = retrieveContext(text)
        val result = llmClient.chat(contextMessages, ragContext = ragContext)

        result.onSuccess { reply ->
            val assistantMsg = ChatMessage(role = ChatMessage.Role.ASSISTANT, content = reply)
            appendToScreen(assistantMsg)
            appendToContext(assistantMsg)
        }.onFailure {
            // Rollback both buffers on failure
            _messages.value = _messages.value.dropLast(1)
            contextBuffer.removeAt(contextBuffer.lastIndex)
        }

        return result
    }

    suspend fun sendStream(text: String): Result<Flow<String>> {
        val userMsg = ChatMessage(role = ChatMessage.Role.USER, content = text)
        appendToScreen(userMsg)
        appendToContext(userMsg)

        val ragContext = retrieveContext(text)

        return try {
            val flow = llmClient.chatStream(contextMessages, ragContext = ragContext)
            Result.success(flow)
        } catch (e: Exception) {
            _messages.value = _messages.value.dropLast(1)
            contextBuffer.removeAt(contextBuffer.lastIndex)
            Result.failure(e)
        }
    }

    fun appendAssistantReply(text: String) {
        if (text.isBlank()) return
        val assistantMsg = ChatMessage(role = ChatMessage.Role.ASSISTANT, content = text)
        appendToScreen(assistantMsg)
        appendToContext(assistantMsg)
    }

    /** Remove the last user message from both screen and context buffer.
     *  Used when a recording is cancelled before the LLM reply arrives,
     *  so the orphaned user message does not affect the next conversation. */
    fun clearLastUserMessage() {
        val screenMsgs = _messages.value.toMutableList()
        if (screenMsgs.isNotEmpty() && screenMsgs.last().role == ChatMessage.Role.USER) {
            screenMsgs.removeAt(screenMsgs.lastIndex)
            _messages.value = screenMsgs
        }
        if (contextBuffer.isNotEmpty() && contextBuffer.last().role == ChatMessage.Role.USER) {
            contextBuffer.removeAt(contextBuffer.lastIndex)
        }
    }

    /** Full clear — both screen and LLM context (user-initiated). */
    fun clear() {
        _messages.value = emptyList()
        contextBuffer.clear()
    }

    val messageCount: Int
        get() = _messages.value.size
}
